features/karaf/esh-core/src/main/feature/feature.xml

```diff
@@ -304,7 +304,22 @@
 
   <feature name="esh-io-transport-serial" version="${project.version}">
```

Comment:
Just for my understanding: This feature can now be regarded as legacy and solutions should always either use esh-io-transport-serial-javacomm or esh-io-transport-serial-rxtx instead, is this correct?
If so, I wouldn't mind to completely remove it - for me there isn't any need for this to be backward compatible.

Reply:
Bundles needs the serial API and "some" implementation.
So, features that require the serial transport depends on `esh-io-transport-serial`.
If no implementation is already installed it falls back to `esh-io-transport-serial-rxtx` and install also this one.
If the dependencies are already satisfied, because a product or user installs `esh-io-transport-serial-javacomm` the dependencies of `esh-io-transport-serial` are already installed and `esh-io-transport-serial-rxtx` is not installed.

We depend on an unspecific `esh-io-transport-serial` feature, that also fallback to a specific implementation if someone is missing.

Have you a better idea how to model the unspecific dependency to "any" implementation?

Reply:
> If the dependencies are already satisfied, because a product or user installs esh-io-transport-serial-javacomm the dependencies of esh-io-transport-serial are already installed

I still haven't fully grasped the Karaf feature dependency management here - sorry, if I ask stupid questions.
But esh-io-transport-serial defines esh-io-transport-serial-rxtx as its dependency, so how can this be already installed, if a user only installed esh-io-transport-serial-javacomm? To me, it appears as if esh-io-transport-serial is tied to rxtx and not being unspecific (while the previous "require feature=serial" looked pretty unspecific).

Reply:
The feature resolution is done using OSGi requirements and capabilities.
All the ones in the manifests of the bundles are taken into account and also that ones that are added to the feature itself.
So, package imports and exports define requirements and capabilities and also the explicit provide capability and require capability ones are taken into account.
As bundles sometimes miss that information it could be added to the feature, too (as it is sometimes not possible to modify the bundles).
We also define a lot of capabilities in the features instead of adding that to the bundles...

If a feature or a dependency is using the attribute `dependency="true"` it will be installed only if there are requirements that are not satisfied without that one and satisfied with that one.

If a bundle requires an OSGi service that implements a specific interface the bundle should contain a specific require capability section.
If a bundle contains an OSGi service that implements a specific interface the bundel should contain a specific provide capability section.
Here you point me with your comment to the situation that our manifest misses the information (e.g. the serialbutton binding) that it needs such a service implementation ... I added the requirement for a specific implementation to the unspecific serial feature now.

So, features itself that needs a serial implementation should depend on "...-serial" that is does not specify the implementation itself. It only brings in one using `dependency="true"` if the capability is not already satisfied.

If an user install before or later a feature that brings in the requirement, the fallback feature can be dropped again, if that feature is removed another time the fallback will be used again...

We already use the dependency attribute for the two different Guice bundles, so the bigger one is only installed if required (and the smaller one could be dropped if the bigger one is installed).

Reply:
Ok, I think I got it. Many thanks for the elaborate explanation!